### ISSUE TITLE
[clang-tidy] bugprone-use-after-move false positive

### ISSUE THREAD

**Issue by Hedede** [llvm/llvm-project#62206]
clang-tidy 15.0.7 produces a warning in the following code on the commented line.
I believe it is a false positive because it is being reassigned.
 
```cpp
#include <memory>
#include <variant>

struct expression;

struct identifier_expression {};

enum class binary_operator {
	add,
	assign,
	assign_add,
};

struct binary_expression {
	binary_operator op;
	std::shared_ptr<expression> lhs;
	std::shared_ptr<expression> rhs;
};

using expression_variant = std::variant<identifier_expression, binary_expression>;
struct expression : expression_variant {
	using expression_variant::expression_variant;
	using expression_variant::operator=;
};

template<typename Value, typename T = std::remove_reference_t<Value>>
auto wrap(Value&& init) -> std::unique_ptr<T>
{
	return std::make_unique<T>(std::forward<Value>(init));
}

auto simplify_binary_expr(binary_expression expr) -> binary_expression
{
	expression lhs = std::move(*expr.lhs);
	expression rhs = std::move(*expr.rhs);
	if (expr.op == binary_operator::assign_add) {
		rhs = binary_expression{ // 'rhs' used after it was moved [bugprone-use-after-move]
			.op = binary_operator::add,
			.lhs = wrap(lhs),
			.rhs = wrap(std::move(rhs)),
		};
		expr.op = binary_operator::assign;
	}

	return {
		.op = expr.op,
		.lhs = wrap(std::move(lhs)),
		.rhs = wrap(std::move(rhs)),
	};
}
```

**Comment by unterumarmung** [comment]
Smaller reproducer: https://godbolt.org/z/M5G8fsenP
